File: backend/app/routers/pricing.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field
from decimal import Decimal
from typing import Optional, List
import numpy as np

router = APIRouter()


# Modelos de datos
class PricingInput(BaseModel):
    product_name: str = Field(
        ..., description="Nombre del producto o servicio"
    )
    cost_materials: float = Field(..., ge=0, description="Costo de materiales")
    cost_labor: float = Field(..., ge=0, description="Costo de mano de obra")
    cost_overhead: float = Field(..., ge=0, description="Gastos generales")
    desired_profit_margin: float = Field(
        ..., ge=0, le=100, description="Margen de ganancia deseado (%)"
    )
    market_research: Optional[dict] = Field(
        None, description="Investigación de mercado"
    )
    competitor_prices: Optional[List[float]] = Field(
        None, description="Precios de competidores"
    )
    target_market: Optional[str] = Field(
        None, description="Mercado objetivo"
    )


class PricingResult(BaseModel):
    recommended_price: float
    cost_breakdown: dict
    profit_margin: float
    competitive_analysis: dict
    pricing_strategies: List[dict]
    financial_projections: dict


class PricingService:
    @staticmethod
    def calculate_base_cost(
        materials: float, labor: float, overhead: float
    ) -> Decimal:
        """Calcula el costo base con precisión decimal"""
        return (
            Decimal(str(materials)) +
            Decimal(str(labor)) +
            Decimal(str(overhead))
        )

    @staticmethod
    def calculate_price_with_margin(
        base_cost: Decimal, margin_percent: float
    ) -> Decimal:
        """Calcula precio con margen deseado"""
        margin_decimal = Decimal(str(margin_percent)) / Decimal('100')
        return base_cost / (Decimal('1') - margin_decimal)

    @staticmethod
    def analyze_competition(competitor_prices: List[float]) -> dict:
        """Analiza precios de competidores"""
        if not competitor_prices:
            return {"message": "No hay datos de competidores"}

        prices = np.array(competitor_prices)
        return {
            "min_price": float(np.min(prices)),
            "max_price": float(np.max(prices)),
            "avg_price": float(np.mean(prices)),
            "median_price": float(np.median(prices)),
            "std_deviation": float(np.std(prices))
        }

    @staticmethod
    def generate_pricing_strategies(
        base_cost: Decimal, competitor_analysis: dict
    ) -> List[dict]:
        """Genera estrategias de precios"""
        strategies = []

        # Estrategia conservadora
        conservative_price = base_cost * Decimal('1.20')  # 20% margen
        strategies.append({
            "name": "Conservadora",
            "price": float(conservative_price),
            "description": "Margen bajo pero seguro, ideal para empezar",
            "pros": ["Fácil venta", "Competitivo"],
            "cons": ["Menor ganancia"]
        })

        # Estrategia competitiva
        if competitor_analysis.get("avg_price"):
            competitive_price = Decimal(str(competitor_analysis["avg_price"]))
            strategies.append({
                "name": "Competitiva",
                "price": float(competitive_price),
                "description": "Precio similar al promedio del mercado",
                "pros": ["Equilibrio", "Aceptación del mercado"],
                "cons": ["Menos diferenciación"]
            })

        # Estrategia premium
        premium_price = base_cost * Decimal('1.50')  # 50% margen
        strategies.append({
            "name": "Premium",
            "price": float(premium_price),
            "description": "Alto margen, enfoque en calidad y valor",
            "pros": ["Mayor ganancia", "Percepción de calidad"],
            "cons": ["Mercado más pequeño"]
        })

        return strategies


@router.post("/calculate", response_model=PricingResult)
async def calculate_pricing(data: PricingInput):
    """Calcula el precio ideal para un producto o servicio"""
    try:
        service = PricingService()

        # Cálculo del costo base
        base_cost = service.calculate_base_cost(
            data.cost_materials,
            data.cost_labor,
            data.cost_overhead
        )

        # Precio con margen deseado
        recommended_price = service.calculate_price_with_margin(
            base_cost,
            data.desired_profit_margin
        )

        # Análisis de competidores
        competitive_analysis = service.analyze_competition(
            data.competitor_prices or []
        )

        # Estrategias de precios
        pricing_strategies = service.generate_pricing_strategies(
            base_cost,
            competitive_analysis
        )

        # Breakdown de costos
        cost_breakdown = {
            "materials": data.cost_materials,
            "labor": data.cost_labor,
            "overhead": data.cost_overhead,
            "total_cost": float(base_cost),
            "profit_amount": float(recommended_price - base_cost)
        }

        # Proyecciones financieras
        financial_projections = {
            "break_even_units": 1,  # Simplificado
            "monthly_revenue_100_units": float(recommended_price * 100),
            "monthly_profit_100_units": float(
                (recommended_price - base_cost) * 100
            ),
            "roi_percentage": float(
                ((recommended_price - base_cost) / base_cost) * 100
            )
        }

        return PricingResult(
            recommended_price=float(recommended_price),
            cost_breakdown=cost_breakdown,
            profit_margin=data.desired_profit_margin,
            competitive_analysis=competitive_analysis,
            pricing_strategies=pricing_strategies,
            financial_projections=financial_projections
        )

    except Exception as e:
        raise HTTPException(
            status_code=400, detail=f"Error en cálculo: {str(e)}"
        )


@router.get("/strategies")
async def get_pricing_strategies():
    """Obtiene información sobre estrategias de precios"""
    return {
        "strategies": [
            {
                "name": "Penetración de Mercado",
                "description": "Precios bajos para ganar participación",
                "best_for": "Productos nuevos, mercados competitivos"
            },
            {
                "name": "Descremado",
                "description": "Precios altos iniciales, luego reducción",
                "best_for": "Productos innovadores, early adopters"
            },
            {
                "name": "Precio Competitivo",
                "description": "Precio similar a competidores",
                "best_for": "Mercados maduros, productos similares"
            },
            {
                "name": "Precio Premium",
                "description": "Precios altos basados en valor percibido",
                "best_for": "Productos de calidad superior, marca fuerte"
            }
        ]
    }

File: src/lib/calculations/pricing.ts
// Librería de cálculos de pricing financiera
export interface PricingInput {
  productName: string;
  costMaterials: number;
  costLabor: number;
  costOverhead: number;
  desiredProfitMargin: number;
  competitorPrices?: number[];
  targetMarket?: string;
}

export interface PricingResult {
  recommendedPrice: number;
  costBreakdown: {
    materials: number;
    labor: number;
    overhead: number;
    totalCost: number;
    profitAmount: number;
  };
  profitMargin: number;
  competitiveAnalysis: {
    minPrice?: number;
    maxPrice?: number;
    avgPrice?: number;
    medianPrice?: number;
    stdDeviation?: number;
    message?: string;
  };
  pricingStrategies: Array<{
    name: string;
    price: number;
    description: string;
    pros: string[];
    cons: string[];
  }>;
  financialProjections: {
    breakEvenUnits: number;
    monthlyRevenue100Units: number;
    monthlyProfit100Units: number;
    roiPercentage: number;
  };
}

export class PricingCalculator {
  /**
   * Calcula el precio ideal con análisis completo
   */
  static calculate(input: PricingInput): PricingResult {
    // Cálculo del costo base
    const totalCost = input.costMaterials + input.costLabor + input.costOverhead;
    
    // Precio con margen deseado
    const marginDecimal = input.desiredProfitMargin / 100;
    const recommendedPrice = totalCost / (1 - marginDecimal);
    
    // Análisis de competidores
    const competitiveAnalysis = this.analyzeCompetition(input.competitorPrices || []);
    
    // Estrategias de precios
    const pricingStrategies = this.generatePricingStrategies(totalCost, competitiveAnalysis);
    
    // Breakdown de costos
    const costBreakdown = {
      materials: input.costMaterials,
      labor: input.costLabor,
      overhead: input.costOverhead,
      totalCost: totalCost,
      profitAmount: recommendedPrice - totalCost
    };
    
    // Proyecciones financieras
    const financialProjections = {
      breakEvenUnits: 1,
      monthlyRevenue100Units: recommendedPrice * 100,
      monthlyProfit100Units: (recommendedPrice - totalCost) * 100,
      roiPercentage: ((recommendedPrice - totalCost) / totalCost) * 100
    };
    
    return {
      recommendedPrice: Math.round(recommendedPrice * 100) / 100,
      costBreakdown,
      profitMargin: input.desiredProfitMargin,
      competitiveAnalysis,
      pricingStrategies,
      financialProjections
    };
  }
  
  /**
   * Analiza precios de competidores
   */
  private static analyzeCompetition(competitorPrices: number[]) {
    if (!competitorPrices || competitorPrices.length === 0) {
      return { message: "No hay datos de competidores" };
    }
    
    const prices = competitorPrices.filter(p => p > 0);
    if (prices.length === 0) {
      return { message: "No hay precios válidos de competidores" };
    }
    
    // Estadísticas básicas
    const minPrice = Math.min(...prices);
    const maxPrice = Math.max(...prices);
    const avgPrice = prices.reduce((a, b) => a + b, 0) / prices.length;
    
    // Mediana
    const sortedPrices = [...prices].sort((a, b) => a - b);
    const medianPrice = sortedPrices.length % 2 === 0
      ? (sortedPrices[sortedPrices.length / 2 - 1] + sortedPrices[sortedPrices.length / 2]) / 2
      : sortedPrices[Math.floor(sortedPrices.length / 2)];
    
    // Desviación estándar
    const variance = prices.reduce((acc, price) => acc + Math.pow(price - avgPrice, 2), 0) / prices.length;
    const stdDeviation = Math.sqrt(variance);
    
    return {
      minPrice: Math.round(minPrice * 100) / 100,
      maxPrice: Math.round(maxPrice * 100) / 100,
      avgPrice: Math.round(avgPrice * 100) / 100,
      medianPrice: Math.round(medianPrice * 100) / 100,
      stdDeviation: Math.round(stdDeviation * 100) / 100
    };
  }
  
  /**
   * Genera estrategias de precios
   */
  private static generatePricingStrategies(baseCost: number, competitorAnalysis: {
    minPrice?: number;
    maxPrice?: number;
    avgPrice?: number;
    medianPrice?: number;
    stdDeviation?: number;
    message?: string;
  }) {
    const strategies = [];
    
    // Estrategia conservadora (20% margen)
    const conservativePrice = baseCost * 1.20;
    strategies.push({
      name: "Conservadora",
      price: Math.round(conservativePrice * 100) / 100,
      description: "Margen bajo pero seguro, ideal para empezar",
      pros: ["Fácil venta", "Competitivo"],
      cons: ["Menor ganancia"]
    });
    
    // Estrategia competitiva
    if (competitorAnalysis.avgPrice) {
      strategies.push({
        name: "Competitiva",
        price: competitorAnalysis.avgPrice,
        description: "Precio similar al promedio del mercado",
        pros: ["Equilibrio", "Aceptación del mercado"],
        cons: ["Menos diferenciación"]
      });
    }
    
    // Estrategia premium (50% margen)
    const premiumPrice = baseCost * 1.50;
    strategies.push({
      name: "Premium",
      price: Math.round(premiumPrice * 100) / 100,
      description: "Alto margen, enfoque en calidad y valor",
      pros: ["Mayor ganancia", "Percepción de calidad"],
      cons: ["Mercado más pequeño"]
    });
    
    return strategies;
  }
}

// Estrategias de pricing adicionales
export const PRICING_STRATEGIES = [
  {
    name: "Penetración de Mercado",
    description: "Precios bajos para ganar participación",
    bestFor: "Productos nuevos, mercados competitivos"
  },
  {
    name: "Descremado",
    description: "Precios altos iniciales, luego reducción",
    bestFor: "Productos innovadores, early adopters"
  },
  {
    name: "Precio Competitivo",
    description: "Precio similar a competidores",
    bestFor: "Mercados maduros, productos similares"
  },
  {
    name: "Precio Premium",
    description: "Precios altos basados en valor percibido",
    bestFor: "Productos de calidad superior, marca fuerte"
  }
];
